### PR TITLE
VTGate - Execute Batch V1

### PR DESCRIPTION
**Pull request by harshit-gangal**
@sougou 
This is an initial PR, contains ExecuteBatch API flow.

TODO (In next Commits): 
- Code Refactoring (Renaming to meaningful names with proper comments)
- Test Cases
- Remaining Implementation

This PR is open for suggestions.

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/2304)

### CODE REVIEW

File: proto/vtgate.proto [hunk 1]
```diff
@@ -252,6 +252,46 @@ message ExecuteEntityIdsResponse {
...
+  Session session = 2;
+
+  // query is the query and bind variables to execute.
+  repeated query.BoundQuery query = 3;
+
+  // tablet_type is the type of tablets that this query is targeted to.
+  topodata.TabletType tablet_type = 4;
+
+  // not_in_transaction is deprecated and should not be used.
+  bool not_in_transaction = 5;
```
Comment (alainjobart) [inline proto/vtgate.proto]: Let's remove this field.

File: proto/vtgate.proto [hunk 2]
```diff
@@ -252,6 +252,46 @@ message ExecuteEntityIdsResponse {
...
+
+  // keyspace to target the query to.
```
Comment (alainjobart) [inline proto/vtgate.proto]: Maybe update this comment:
// default keyspace to target the queries to.

File: proto/vtgate.proto [hunk 3]
```diff
@@ -252,6 +252,46 @@ message ExecuteEntityIdsResponse {
...
+  string keyspace = 6;
+
+  // options
+  query.ExecuteOptions options = 7;
+}
```
Comment (alainjobart) [inline proto/vtgate.proto]: How about adding as_transaction in there too? Even if we don't support it in all cases just yet...

File: proto/vtgate.proto [hunk 4]
```diff
@@ -252,6 +252,46 @@ message ExecuteEntityIdsResponse {
...
+
+// ExecuteBatchResponse is the returned value from ExecuteBatch.
+message ExecuteBatchResponse {
+  // error contains an application level error if necessary. Note the
+  // session may have changed, even when an error is returned (for
+  // instance if a database integrity error happened).
+  repeated vtrpc.RPCError error = 1;
```
Comment (alainjobart) [inline proto/vtgate.proto]: I think this field and the result should be combined in a new structure, and this should contain an array of it. And it should probably be defined in query.proto, as you'll need it there for the vttablet API anyway.

To be clearer, in query.proto, define:

message QueryResultOrError {
  vtrpc.RPCError error = 1;
  QueryResult result = 2;
}
(that message name is not very good!)

And this should be:

message ExecuteBatchResponse {
  repeated query.QueryResultOrError results = 1;
  Session session = 2;
}

**Comment by harshit-gangal** [comment]
What BeginExecuteBatch does?

https://github.com/youtube/vitess/blob/master/proto/query.proto#L454

**Comment by harshit-gangal** [comment]
Review status: 0 of 25 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[proto/vtgate.proto, line 272 at r1](https://reviewable.io:443/reviews/youtube/vitess/2304#-KZ2iKw2fAe-riNgrgJw-r1-272:-KZ2j15XqDCCPELJweTW:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/92e9850dc07523432509b48370f2a3501563c725/proto/vtgate.proto#L272)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

Let's remove this field.
</blockquote></details>

Done.

---

*[proto/vtgate.proto, line 274 at r1](https://reviewable.io:443/reviews/youtube/vitess/2304#-KZ2iKw2fAe-riNgrgJw-r1-274:-KZ2kROt-NQynFynk57C:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/92e9850dc07523432509b48370f2a3501563c725/proto/vtgate.proto#L274)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

Maybe update this comment:
// default keyspace to target the queries to.
</blockquote></details>

Done.

---

*[proto/vtgate.proto, line 279 at r1](https://reviewable.io:443/reviews/youtube/vitess/2304#-KZ2iKw2fAe-riNgrgJw-r1-279:-KZ2jUbQss7jjmTm_7Bj:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/92e9850dc07523432509b48370f2a3501563c725/proto/vtgate.proto#L279)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

How about adding as_transaction in there too? Even if we don't support it in all cases just yet...
</blockquote></details>

Done.

---

*[proto/vtgate.proto, line 286 at r1](https://reviewable.io:443/reviews/youtube/vitess/2304#-KZ2iKw2fAe-riNgrgJw-r1-286:-KZ2jdDBPKmzODZfY3hY:b-wnd1w1) ([raw file](https://github.com/youtube/vitess/blob/92e9850dc07523432509b48370f2a3501563c725/proto/vtgate.proto#L286)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

I think this field and the result should be combined in a new structure, and this should contain an array of it. And it should probably be defined in query.proto, as you'll need it there for the vttablet API anyway.

To be clearer, in query.proto, define:

message QueryResultOrError {
  vtrpc.RPCError error = 1;
  QueryResult result = 2;
}
(that message name is not very good!)

And this should be:

message ExecuteBatchResponse {
  repeated query.QueryResultOrError results = 1;
  Session session = 2;
}

</blockquote></details>

We can do vttablet side changes in different PR as those changes are not part of version 1.
Also, we agreed that we will add execute_all flag in the existing ExecuteBatch in query.proto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2304)*

**Comment by harshit-gangal** [comment]
Gentle Reminder..

File: go/vt/vtgate/vtgate.go [hunk 5]
```diff
@@ -353,6 +353,62 @@ func (vtg *VTGate) ExecuteEntityIds(ctx context.Context, sql string, bindVariabl
...
+			"AsTransaction":    asTransaction,
+			"Options":          options,
+		}
+		if bindVariables != nil {
+			query["BindVariables"] = bindVariables[row]
+		}
+		errs = append(errs, handleExecuteError(err[row], statsKey, query, vtg.logExecute))
+	}
+
+	return nil, errs*/
```
Comment (sougou) [inline go/vt/vtgate/vtgate.go]: Delete commented out section.

File: proto/vtgate.proto [hunk 6]
```diff
@@ -256,6 +256,60 @@ message ExecuteEntityIdsResponse {
...
+  string keyspace = 6;
+
+  // options
+  query.ExecuteOptions options = 7;
+}
+
+// QueryResponse represents a query response
+// in the form of result or error but not both.
+// This is used in a list for ExecuteBatchResponse.
+message QueryResponse {
```
Comment (sougou) [inline proto/vtgate.proto]: This name was too ambiguous. Let's go with `ResultWithError` instead.

File: proto/vtgate.proto [hunk 7]
```diff
@@ -256,6 +256,60 @@ message ExecuteEntityIdsResponse {
   query.QueryResult result = 3;
 }
 
+// ExecuteBatchRequest is the payload to ExecuteBatch.
+message ExecuteBatchRequest {
```
Comment (sougou) [inline proto/vtgate.proto]: Need to move these to query.proto.

**Comment by harshit-gangal** [comment]
Review status: 0 of 28 files reviewed at latest revision, 7 unresolved discussions.

---

*[go/vt/vtgate/vtgate.go, line 384 at r11](https://reviewable.io:443/reviews/youtube/vitess/2304#-KZ2iKw1YxZ0W2YkOdWn-r11-384:-KZPv5djut55HJPmy8RL:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/37023b4a906335139d9d9d442d83b6d1d6dfbe3f/go/vt/vtgate/vtgate.go#L384)):*
<details><summary><i>Previously, sougou wrote…</i></summary><blockquote>

Delete commented out section.
</blockquote></details>

Done.

---

*[proto/vtgate.proto, line 260 at r11](https://reviewable.io:443/reviews/youtube/vitess/2304#-KZ2iKw2fAe-riNgrgJw-r11-260:-KZPv6tCoQOrcb0dbz2a:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/37023b4a906335139d9d9d442d83b6d1d6dfbe3f/proto/vtgate.proto#L260)):*
<details><summary><i>Previously, alainjobart (Alain Jobart) wrote…</i></summary><blockquote>

No sorry, it's ResultWithError that should be in query.proto, ExecuteBatchRequest and friends should stay in vtgateservice :)
</blockquote></details>

Done.

---

*[proto/vtgate.proto, line 291 at r11](https://reviewable.io:443/reviews/youtube/vitess/2304#-KZ2iKw2fAe-riNgrgJw-r11-291:-KZPv8P2cYCWzQCyNo27:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/37023b4a906335139d9d9d442d83b6d1d6dfbe3f/proto/vtgate.proto#L291)):*
<details><summary><i>Previously, sougou wrote…</i></summary><blockquote>

This name was too ambiguous. Let's go with `ResultWithError` instead.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2304)*

**Comment by sougou** [comment]
Looks like you have to resolve a merge conflict after all.

**Comment by sougou** [comment]
LGTM

**Pull request merged**